# Assignment - A: Build a Regression Model in Keras

## Download and Clean Dataset

Let's start by importing the pandas and the Numpy libraries.

In [36]:
import pandas as pd
import numpy as np

We will be using the same dataset provided in the course.

The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:

1. Cement
2. Blast Furnace Slag
3. Fly Ash
4. Water
5. Superplasticizer
6. Coarse Aggregate
7. Fine Aggregate

Let's read the dataset into a *pandas* dataframe.

In [37]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 

Let's check how many data points we have.

In [38]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.

Let's check the dataset for any missing values.

In [39]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [40]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

### Split data into predictors and target

The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [41]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

Let's do a quick sanity check of the predictors and the target dataframes.

In [42]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [43]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Let's save the number of predictors to n_cols since we will need this number when building our network.

In [44]:
n_cols = predictors.shape[1] # number of predictors
n_cols

8

## Build a Baseline Model

### Import Keras

The extremely powerful and yet-easy-to-use **Keras** library runs on top of a low-level library such as **TensorFlow** or **PyTorch**. In this case, this assignment will be using **TensorFlow** as a backend.

In [45]:
import keras

As you can see, the TensorFlow backend was used to install the Keras library.

Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [46]:
from keras.models import Sequential
from keras.layers import Dense

### Build a Neural Network

Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.

In [47]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function create a model using the following settings:
- 1 hidden layer of *10 nodes*
- *ReLU* activation function
- *adam* optimizer
- *mean squared error* for the loss function

### Generate the Test Dataset

Let's import scikit-learn in order to randomly split the data into a training and test sets

In [48]:
from sklearn.model_selection import train_test_split

Randomly split the data into a training and test sets by holding 30% of the data for testing.

In [49]:
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)

### Train and Evaluate the Neural Network Model

Let's call the function now to create our model.

In [50]:
# build the model
model = regression_model()

Next, we will train the model for 50 epochs with 1 verbose setting.

In [51]:
# fit the model - 50 epochs with 1 verbose setting
model.fit(X_train, y_train, epochs=50, verbose=1)

Epoch 1/50
721/721 [==============================] - 2s 3ms/step - loss: 2198.5144
Epoch 2/50
721/721 [==============================] - 2s 2ms/step - loss: 1520.0501
Epoch 3/50
721/721 [==============================] - 2s 2ms/step - loss: 1216.8315
Epoch 4/50
721/721 [==============================] - 2s 3ms/step - loss: 989.4396
Epoch 5/50
721/721 [==============================] - 2s 3ms/step - loss: 818.1559
Epoch 6/50
721/721 [==============================] - 2s 3ms/step - loss: 653.0682
Epoch 7/50
721/721 [==============================] - 2s 3ms/step - loss: 526.5008
Epoch 8/50
721/721 [==============================] - 2s 2ms/step - loss: 424.1948
Epoch 9/50
721/721 [==============================] - 2s 2ms/step - loss: 355.7411
Epoch 10/50
721/721 [==============================] - 2s 3ms/step - loss: 303.2754A: 0s - loss: 309.10
Epoch 11/50
721/721 [==============================] - 2s 2ms/step - loss: 260.7527
Epoch 12/50
721/721 [==============================] - 2s 2ms/

Next we need to evaluate the model on the test data.

In [52]:
# evaluate the model using the test dataset
loss_val = model.evaluate(X_test, y_test)

# varify the prediction
y_pred = model.predict(X_test)

# output the evaluate value
print(loss_val)

# output the prediction
# print(y_pred)

309/309 [==============================] - 0s 2ms/step
122.88544748053195


Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.

Let's import the *mean_squared_error* function from *scikit-learn*.

In [53]:
from sklearn.metrics import mean_squared_error

Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

In [54]:
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)

print(mean, standard_deviation)

122.88544739661614 0.0


Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [55]:
total_mean_squared_errors = 50
epochs = 50

mean_squared_errors = []

for i in range(0, total_mean_squared_errors):
    # split the train and test datasets
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
    
    # fit the model
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    
    # evaluate the model
    mse = model.evaluate(X_test, y_test, verbose=0)
    print("mean_square_error at " + str(i+1) + " is " + str(mse))
    
    # get the prediction
    y_pred = model.predict(X_test)
    
    # get the mean square error
    mean_square_error = mean_squared_error(y_test, y_pred)
    
    # append result to the mean_squared_errors list
    mean_squared_errors.append(mean_square_error)

# convert to array list
mean_squared_errors = np.array(mean_squared_errors)

# get mean value from the array list
mean = np.mean(mean_squared_errors)

# get standard deviation using the array list
standard_deviation = np.std(mean_squared_errors)

print('\n')
print('\n')
print("The output below is the mean and standard deviation of " + str(total_mean_squared_errors) + " mean squared errors without normalized data.")
print("- Number of epochs for each training: " + str(epochs))
print("- Mean: " + str(mean))
print("- Standard Deviation: " + str(standard_deviation))

mean_square_error at 1 is 85.17833117142465
mean_square_error at 2 is 88.50628904077227
mean_square_error at 3 is 67.9527745786994
mean_square_error at 4 is 72.85181935628255
mean_square_error at 5 is 66.69403952688076
mean_square_error at 6 is 64.16202221719192
mean_square_error at 7 is 70.75473461953567
mean_square_error at 8 is 61.93180886672924
mean_square_error at 9 is 58.60612006326323
mean_square_error at 10 is 54.97143023839661
mean_square_error at 11 is 61.655313362195656
mean_square_error at 12 is 48.341843490847495
mean_square_error at 13 is 56.04989104903632
mean_square_error at 14 is 62.07149295745158
mean_square_error at 15 is 50.01573841625819
mean_square_error at 16 is 51.084679081987794
mean_square_error at 17 is 51.45828755696615
mean_square_error at 18 is 50.389116342785286
mean_square_error at 19 is 43.876153964440796
mean_square_error at 20 is 50.3311249199037
mean_square_error at 21 is 43.621328113148515
mean_square_error at 22 is 45.23569735431362
mean_square_err